### PR TITLE
Improve python client to support StudioAPI security and authentication features

### PR DESCRIPTION
**Pull request by martlaak**
* fix older tests failing since the _handle_container_message() API change

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -54,16 +54,51 @@ Client(host, port, auto_reconnect)
 
     auto_reconnect - Optional argument to enable/disable automatic reconnect when connection is lost. Defaults to True if not specified.
 
+    user_id - Optional argument to use when server requires authentication - user id (username)
+
+    password  - Optional argument to use when server requires authentication - user password
+
+    use_encryption - Optional boolean argument to set when server uses encryption
```
Comment (Karmo7) [inline README.rst]: set when server uses encryption? Couldn't the client somehow detect that automatically? How does Studio manage to do it?

Comment (Karmo7) [inline README.rst]: Also if it is optional, then what is the default?

File: cdp_client/cdp.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from promise import Promise
 from time import sleep
 from collections import namedtuple
+from hashlib import sha256
```
Comment (nuustik) [inline cdp_client/cdp.py:4]: is this part of standard python? if not then dependencies go to setup.py to be downloaded automatically

File: .gitignore [hunk 3]
```diff
@@ -1,7 +1,8 @@
 *.pyc
 /dist/
+/venv/
```
Comment (nuustik) [inline .gitignore:3]: you also need to bump up the version once changes are applied, and keep in mind that every change requires a version change once you have uploaded a new package. so make sure your changes are final before pushing new version to pypi.org, which must also be done

File: README.rst [hunk 4]
```diff
@@ -54,16 +54,51 @@ Client(host, port, auto_reconnect)
...
+
+        client = cdp.Client('127.0.0.1', user_id="test", password="12345678")
+
+- Usage example with password authentication and encryption in use, without server certificate verification
+
+    .. code:: python
+
+        import ssl
+
+        client = cdp.Client('127.0.0.1', user_id="test", password="12345678", use_encryption=True,
```
Comment (nuustik) [inline README.rst]: for consistency first parameter should also be 
host="127.0.0.1" if possible

File: README.rst [hunk 5]
```diff
@@ -54,16 +54,51 @@ Client(host, port, auto_reconnect)
...
+                                         encryption_parameters={"cert_reqs": ssl.CERT_NONE})
+
+
+- Usage example with password authentication and encryption in use, with server certificate verification
+
+    .. code:: python
+
+        import ssl
+
+        client = cdp.Client('127.0.0.1', user_id="test", password="12345678", use_encryption=True,
```
Comment (nuustik) [inline README.rst]: also i think we have used single quotes around strings, should be consistent

File: cdp_client/cdp.py [hunk 6]
```diff
@@ -48,8 +49,9 @@ class UnknownError(Exception):
 
 
 class Client:
-    def __init__(self, host, port=7689, auto_reconnect=True):
-        self._connection = Connection(host, port, auto_reconnect)
+    def __init__(self, host, port=7689, auto_reconnect=True, use_encryption=False, encryption_parameters=dict(),
```
Comment (nuustik) [inline cdp_client/cdp.py]: probably host should also have default '127.0.0.1'

**Review by nuustik — APPROVED** [review]

File: README.rst (hunk 1 above)
Comment (martlaak) [reply README.rst]: IMO this is not needed in client. Just makes client more complex. Logic is similar to the port number that is also not auto-detected in the python/JS clients. And you probably anyway need to set up the certificates verification to work it securely etc.

File: cdp_client/cdp.py (hunk 2 above)
Comment (martlaak) [reply cdp_client/cdp.py:4]: IMO it is standard pyhton?

File: README.rst (hunk 1 above)
Comment (martlaak) [reply README.rst]: Added note that this is False

File: cdp_client/cdp.py (hunk 2 above)
Comment (nuustik) [reply cdp_client/cdp.py:4]: depends if you installed it with pip install or not, maybe it is

File: .gitignore (hunk 3 above)
Comment (martlaak) [reply .gitignore:3]: Changed version='1.1.1' to version='1.2' in setup.py

File: README.rst (hunk 4 above)
Comment (martlaak) [reply README.rst]: Done

File: cdp_client/cdp.py (hunk 6 above)
Comment (martlaak) [reply cdp_client/cdp.py]: Added the default 127.0.01

File: cdp_client/cdp.py (hunk 2 above)
Comment (martlaak) [reply cdp_client/cdp.py:4]: I haven't installed it intentionally. Neither list my `pip list installed` it.

File: README.rst (hunk 5 above)
Comment (martlaak) [reply README.rst]: Changed to single quotes

File: README.rst (hunk 1 above)
Comment (Karmo7) [reply README.rst]: It is just a weird boolean, I have not seem similar things in REST APIs where one needs to know whether it is http or https.
Maybe instead the first parameter should include it (something like 'wss://127.0.0.1')? And if protocol is missing, we presume no encryption for backwards compatibility

Comment (martlaak) [reply README.rst]: IMO `use_encryption` is quite user-friendly and should be very understandable that you need to add it to get connected to CDP system that also has "Use Encryption" flag set. 
At least to me it seems more user-friendly, than to explain that you have to add "wss://" to the host when you need to connect to CDP system that has "Use Encryption" flag set.

Comment (Karmo7) [reply README.rst]: The `use_encryption` is a bit ambiguous. At least my first guess was that the client chooses whether it wants that one specific connection to be encrypted. But actually it is the server that made the choice, so only one value is correct. Therefore the boolean is more like `is_server_using_encryption` which I admit is a bit wordy. I don't really have a good idea here,  just hoped handling it same way as http/https choice would feel more familiar to users.

Seems connecting with the internal websocket library we use also does it with protocol 
```self._ws = self._connect(protocol + host + ":" + str(port))```

Comment (martlaak) [reply README.rst]: I chose `use_encryption` to be similar/familiar to the "Use Encryption" setting in server. And fact is client has to "use" encryption on connect :) Yes client can not force encryption but it has to use it - otherwise connection would not succeed. Prefixing host with "wss://" is just another, more complicated way of stating that client has to use encryption. I bet many of our automation background users do not know anything about ws:// or wss:// protocol prefixes or what they stand for or how they differ.

**Review by Karmo7 — APPROVED** [review]

File: README.rst (hunk 1 above)
Comment (Karmo7) [reply README.rst]: Ok, next question. If I get the `use_encryption` wrong, do I get a friendly error message explaining what I must do?

Because I might not know whether it is supposed to be encrypted or not if somebody else deployed the system. Maybe CDP should print out "StudioAPIServer: listening on wss://192.168.0.18:7690" or "... Encryption is enabled"?

Comment (martlaak) [reply README.rst]: StudioAPIServer is currently printing "StudioAPIServer: listening on 127.0.0.1:7689 (TLS)" - we should maybe state something more user-friendly instead/in addition of `(TLS)`
And maybe we should improve our JS/python clients - they are currently not user-friendly at all when you misconfigure them - either port is wrong or host is wrong or authentication errors or encryption problems (e.g. certificate problems) - we should decide how and where to show errors and are they customizable or not.
